### PR TITLE
Implement AddPositionalEncoding transform

### PR DESCRIPTION
**Pull request by dongkwan-kim**
This PR includes implementations of Laplacian eigenvector positional encoding & random walk positional encoding based on the original authors' implementation.

- https://github.com/graphdeeplearning/benchmarking-gnns/blob/ef8bd8c7d2c87948bc1bdd44099a52036e715cd0/data/COLLAB.py#L12
- https://github.com/vijaydwivedi75/gnn-lspe/blob/21d2a654569fcac42cde28ce0695386d1c56a9b9/data/ogb_mol.py#L97

Any comments are welcome. Plus, it looks fine to me but I do not know why there is an indentation error in the documentation build...

### CODE REVIEW

File: torch_geometric/transforms/add_positional_encoding.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            Structural and Positional Representations"
+            <https://arxiv.org/abs/2110.07875>`_ paper.
+        num_channels: (int): The number of channels. If :attr:`name` is
+            :obj:`"laplacian_eigenvector_pe"`, it will be the number of the
+            smallest non-trivial eigenvectors. If :attr:`name` is
+            :obj:`"random_walk_pe"`, it will be the number of the random walk
+            steps.
+        method: (str, optional): The method to add positional encoding.
+            If set to :obj:`"attr"`, will be added to the data object
+                as a new attribute.
```
Comment (Padarn) [inline torch_geometric/transforms/add_positional_encoding.py:42]: I guess this is the doc building error ( also on line 44)

**Comment by Padarn** [comment]
I guess the paper is this one: [GRAPH NEURAL NETWORKS WITH LEARNABLE
STRUCTURAL AND POSITIONAL REPRESENTATIONS](https://arxiv.org/pdf/2110.07875.pdf)? Perhaps we can add a reference to it in the doc?

File: torch_geometric/transforms/add_positional_encoding.py (hunk 1 above)
Comment (dongkwan-kim) [reply torch_geometric/transforms/add_positional_encoding.py]: Thank you for your help. I just fixed it.

**Comment by dongkwan-kim** [comment]
> I guess the paper is this one: [GRAPH NEURAL NETWORKS WITH LEARNABLE STRUCTURAL AND POSITIONAL REPRESENTATIONS](https://arxiv.org/pdf/2110.07875.pdf)? Perhaps we can add a reference to it in the doc?

@Padarn  I added the paper you mentioned in the docs of lines 32 -- 34.
https://github.com/pyg-team/pytorch_geometric/blob/9f52a534c8b15f88e00ecb13462a365c47e937b7/torch_geometric/transforms/add_positional_encoding.py#L32-L34

Is this what you mean, or if there is anything I missed, could you specify it? Thank you.

**Comment by Padarn** [comment]
Yes thats all I meant, just common practice across the docstrings, thanks!

File: torch_geometric/transforms/add_positional_encoding.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            computation of the eigenvectors. (default: :obj:`False`)
+    """
+    def __init__(self, name: str, num_channels: int,
+                 method: Optional[str] = 'attr',
+                 is_undirected: Optional[bool] = False):
+        self.name = name
+        self.num_channels = num_channels
+        self.method = method
+        self.is_undirected = is_undirected
+        assert name in ['laplacian_eigenvector_pe', 'random_walk_pe']
```
[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply torch_geometric/transforms/add_positional_encoding.py:68]: That's a good idea. I will work on it soon.

File: torch_geometric/transforms/add_positional_encoding.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                                                    normalization='sym',
+                                                    dtype=torch.float,
+                                                    num_nodes=N)
+            L = to_scipy_sparse_matrix(edge_index, edge_weight, N)
+            eig_vals, eig_vecs = eig_fn(L, k=self.num_channels + 1, which='SR',
+                                        return_eigenvectors=True)
+            eig_vecs = np.real(eig_vecs[:, eig_vals.argsort()])
+            pe = torch.from_numpy(eig_vecs[:, 1:self.num_channels + 1])
+
+            sign = -1. + 2 * torch.randint(0, 2, (self.num_channels, ),
```
[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply torch_geometric/transforms/add_positional_encoding.py:86]: That's because either positive or negative eigenvectors are both valid. To resolve this sign ambiguity, the original authors propose a random sign flip during training steps. Line 74 is implementing this procedure (multiplying -1 or 1 to eigenvectors). See page 7 in https://arxiv.org/abs/2003.00982 for more details.

[the remark this answers is not in the document]
Comment (Padarn) [reply torch_geometric/transforms/add_positional_encoding.py:86]: ahh okay interesting - thanks for explaining and the reference

File: torch_geometric/transforms/add_positional_encoding.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+from torch_geometric.transforms import BaseTransform
+from torch_geometric.utils import (
+    get_laplacian,
+    remove_self_loops,
+    to_scipy_sparse_matrix,
+)
+from torch_geometric.utils.num_nodes import maybe_num_nodes
+
+
+def diagonal_weight(edge_index, edge_weight, num_nodes=None):
```
Comment (rusty1s) [inline torch_geometric/transforms/add_positional_encoding.py:21]: Worth adding to `torch_geometric.utils.loop`?

File: torch_geometric/transforms/add_positional_encoding.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    loop_index = edge_index[:, loop_mask][0]
+    loop_edge_attr = edge_weight[loop_mask]
+
+    N = maybe_num_nodes(edge_index, num_nodes)
+    diag_attr = torch.zeros((N, ), dtype=edge_weight.dtype)
+    diag_attr[loop_index] = loop_edge_attr
+    return diag_attr
+
+
+class BaseAddPE(BaseTransform):
```
Comment (rusty1s) [inline torch_geometric/transforms/add_positional_encoding.py:32]: Not really a big fan of the base class here. For now, I suggest to make `add_pe` a function outside the class:
```python
def add_attr(data: Data, value: Any, cat: bool, attr_name: Optional[str] = None):
```
We could also think of adding this function to `base_transform.py`, and use it in other `transforms` as well, e.g., in `transforms.Cartesian`. WDYT?

**Review by rusty1s — APPROVED** [review]
This looks pretty good. Thank you very much! Only have some minor points regarding code structure.

File: torch_geometric/transforms/add_positional_encoding.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+
+@functional_transform('add_laplacian_eigenvector_pe')
+class AddLaplacianEigenvectorPE(BaseAddPE):
+    r"""Adds Laplacian eigenvector positional encoding from the `"Benchmarking
+    Graph Neural Networks" <https://arxiv.org/abs/2003.00982>`_ paper to the
+    given graph (functional name: :obj:`add_laplacian_eigenvector_pe`).
+
+    Args:
+        num_channels: (int): The number of channels. It will be the number of
```
[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply torch_geometric/transforms/add_positional_encoding.py:64]: I will remove all `:` after arg names in other documentation. In fact, this was copied and pasted from https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/transforms/add_self_loops.py#L17.

Is the `:` in add_self_loops.py intentional, or just a mistake?

File: torch_geometric/transforms/add_positional_encoding.py (hunk 4 above)
Comment (dongkwan-kim) [reply torch_geometric/transforms/add_positional_encoding.py:21]: That is a good idea. I will work on it first.

File: torch_geometric/transforms/add_positional_encoding.py (hunk 5 above)
Comment (dongkwan-kim) [reply torch_geometric/transforms/add_positional_encoding.py:32]: This is actually a fantastic idea. However, it might modify many files (although I have not counted yet). So, how about creating `add_attr` outside the class for now, and making it a part of `base_transform.py` later in the other PR?

Comment (Padarn) [reply torch_geometric/transforms/add_positional_encoding.py:32]: Maybe it could be part of `Data`?

File: torch_geometric/transforms/add_positional_encoding.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/transforms/add_positional_encoding.py]: Yes, this is a mistake.

[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply torch_geometric/transforms/add_positional_encoding.py]: I see. And I just found that there are many `:` mistakes in other files. I will correct them later.

File: test/transforms/test_add_positional_encoding.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.transforms import (
+    AddLaplacianEigenvectorPE,
+    AddRandomWalkPE,
+)
+
+
+def test_add_laplacian_eigenvector_pe():
```
[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply test/transforms/test_add_positional_encoding.py:10]: I added the output tests for RandomWalkPE.
However, for LaplacianEigenvectorPE, I found that each run produces different outputs. I guess computing eigenvectors of Laplacian is not a deterministic operation. How can we write output tests for LaplacianEigenvectorPE? Any suggestion?

[the remark this answers is not in the document]
Comment (Padarn) [reply test/transforms/test_add_positional_encoding.py:10]: You can try setting the random seed of torch and numpy?

[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply test/transforms/test_add_positional_encoding.py:10]: I tried `np.random.seed` and `np.random.seed`, but it still creates different outputs. I found that using a fixed value for `v0` in `scipy.linalg.eigs` produces deterministic outputs. We can place **kwargs in `__call__` and pass the kwargs (including `v0`) into eigs explicitly. What do you think?

[the remark this answers is not in the document]
Comment (Padarn) [reply test/transforms/test_add_positional_encoding.py:10]: Yeah, it looks like when it's not provided there is a flag 'info=0' [set](https://github.com/scipy/scipy/blob/f2ec91c4908f9d67b5445fbfacce7f47518b35d1/scipy/sparse/linalg/eigen/arpack/arpack.py#L323). 

How different are the outputs, personally I feel maybe adding some leniency in the checking of the result (i.e using `all_close` with a high `atol` pr `rtol`) might be better than adding the extra `**kwargs` just for this.

Just my opinion though. I think both ways work.

[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply test/transforms/test_add_positional_encoding.py:12]: I have tested various graphs for `AddLaplacianEigenvectorPE`. If there is no `v0`, the output really varies so we cannot test with `torch.allclose`.

Instead, I added a clustering test on a graph with two clusters, which uses the mean value of the first non-trivial eigenvector ([Fiedler vector](https://en.wikipedia.org/wiki/Algebraic_connectivity)) of each cluster.

Also, I added tests with exact values using `seed_everything` and `v0`. The keyword argument `v0` will be added through `AddLaplacianEigenvectorPE.__init__`.

[the remark this answers is not in the document]
Comment (dongkwan-kim) [reply test/transforms/test_add_positional_encoding.py:12]: Well... But looking at failed tests, using `seed_everything` and `v0` does not guarantee the same output in different machines.

**Comment by dongkwan-kim** [comment]
This is super great. Thank you all. I will send other PRs mentioned in this thread soon.

**Comment by rusty1s** [comment]
Thank you!

File: test/transforms/test_add_positional_encoding.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Padarn) [reply test/transforms/test_add_positional_encoding.py:12]: Got it. I think what you've done for the clustering test is a good solution. Thanks!

**Comment by Padarn** [comment]
Great!
